minimal file contraction_mapping

agjftucker opened the pull request:
Rohan Mitta seems to have disappeared and I must do the same tomorrow, so I don't think we'll be able to synthesise our work. In any case I'm afraid I have diverged somewhat from his approach (the usual one) by following the approach free of geometric sums due to Palais. So here is a minimal "alternative" pull request containing only my contribution, leading to the main theorem but without all the extras from Sutherland. Alistair.

TO CONTRIBUTORS:

Make sure you have:

 * [x] reviewed and applied the coding style: [coding](./docs/style.md), [naming](./docs/naming.md)
 * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](./docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](./tests/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

For reviewers: [code review check list](./docs/code-review.md)

agjftucker committed:
minimal file contraction_mapping

Approach free of geometric sums, due to Palais.

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+      (mul_comm (dist x y) K ▸ mul_lt_of_lt_div (lt_of_lt_of_le hε h) hd))⟩))
+
+lemma iterated_lipschitz_of_lipschitz {α : Type*} [metric_space α] {K : ℝ} {f : α → α} :
+   lipschitz K f → ∀ (n : ℕ), lipschitz (K ^n) (f ^[n]) :=
```
Suggested whitspace:

```suggestion
   lipschitz K f → ∀ (n : ℕ), lipschitz (K ^ n) (f^[n]) :=
```

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+  { exact h n.2, },
+end
+
+section prod
```
This section should be in `algebra.pi_instance`!

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+section prod
+  open lattice
+
+  local attribute [instance]
```
this will not work in Lean 4, change to

```lean
def prod_has_le ...

local attribute [instance] prod_has_le
```

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+    sup_le        := λ k m n h₁ h₂, ⟨sup_le h₁.1 h₂.1, sup_le h₁.2 h₂.2⟩,
+    .. prod_partial_order}
+
+  lemma prod_at_top_at_top_eq {β₁ β₂ : Type*} [inhabited β₁] [inhabited β₂] [semilattice_sup β₁]
```
move to `order.filter`

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+    (λ h, let ⟨N, hN⟩ := mem_at_top_sets.mp h in mem_prod_iff.mpr
+      ⟨{n₁ | N.1 ≤ n₁}, mem_at_top N.1, {n₂ | N.2 ≤ n₂}, mem_at_top N.2, (λ n hn, hN n hn)⟩))
+
+  lemma prod_map_def {α₁ α₂ β₁ β₂ : Type*} {u₁ : β₁ → α₁} {u₂ : β₂ → α₂} :
```
move to `data.prod`

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+    prod.map u₁ u₂ = λ (n : β₁ × β₂), (u₁ n.1, u₂ n.2) :=
+  funext (λ n, prod.ext (prod.map_fst u₁ u₂ n) (prod.map_snd u₁ u₂ n))
+
+  lemma prod_filter_map_at_top {α₁ α₂ β₁ β₂ : Type*} [inhabited β₁] [inhabited β₂]
```
move to `order.filter`

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+    filter.prod (map u₁ at_top) (map u₂ at_top) = map (prod.map u₁ u₂) at_top :=
+  by rw [prod_map_map_eq, prod_at_top_at_top_eq, prod_map_def]
+
+  lemma prod_dist_eq {α β₁ β₂ : Type*} [metric_space α] (u₁ : β₁ → α) (u₂ : β₂ → α) (n : β₁ × β₂) :
```
move to `analysis.metric_space`

johoelzl commented on analysis/contraction_mapping.lean:
```diff
@@ -0,0 +1,208 @@
+    dist (prod.map u₁ u₂ n).1 (prod.map u₁ u₂ n).2 = dist (dist (u₁ n.1) (u₂ n.2)) 0 :=
+  by rw [prod.map_fst, prod.map_snd, real.dist_0_eq_abs, abs_of_nonneg dist_nonneg]
+
+  lemma cauchy_seq_iff {α β : Type*} [uniform_space α] [inhabited β] [semilattice_sup β]
```
move to `analysis.uniform_space`

johoelzl reviewed (changes requested):
Looks good. Just move some theorems to the correct place

rwbarton commented on analysis/topology/continuity.lean:
```diff
@@ -774,6 +774,18 @@ instance [compact_space α] [compact_space β] : compact_space (α × β) :=
 
+lemma continuous_prod_fst {f : α × β → γ} {b : β} (hf : continuous f) :
+  continuous (λ a, f (a, b)) :=
+λ s hs, subset_interior_iff_open.mp (λ a hm,
```
This can be composed from other continuity lemmas as
```lean
(continuous.prod_mk continuous_id continuous_const).comp hf
```
It may be too specialized to merit its own lemma, though I could see a case for keeping a version with `f` removed (proving `continuous (λ a, (a, b))`).

Commit:
Scatter lemmas among the files

I don't seem to be able to get rid of continuous_prod_snd right now but I have replaced its proof
with @rwbarton's suggestion and moved it back to contraction_mapping.lean.
